File: scripts/export-jsonl.js
/* Export JSONL (one JSON per line) with .json suffix for console import */
const fs = require('node:fs')
const path = require('node:path')
let XLSX
try { XLSX = require('xlsx') } catch (e) { XLSX = require(path.resolve(__dirname, '../functions/import-xlsx/node_modules/xlsx')) }

const SRC = process.argv[2] || path.resolve(process.cwd(), 'prepare/b.xlsx')
const OUT_DIR = process.argv[3] || path.resolve(process.cwd(), 'export')

const COLS = {
  name: ['姓名'],
  id_card: ['身份证号'],
  checkInDate: ['入住时间'],
  birthDate: ['出生日期'],
  gender: ['性别'],
  nativePlace: ['籍贯'],
  ethnicity: ['民族'],
  hospital: ['就诊医院'],
  hospitalDiagnosis: ['医院诊断'],
  doctorName: ['医生姓名'],
  symptoms: ['症状详情'],
  medicalCourse: ['医治过程', '医疗过程'],
  followupPlan: ['后续治疗安排'],
  fatherMixed: ['父亲姓名、电话、身份证号'],
  motherMixed: ['母亲姓名、电话、身份证号', '母亲姓名、电话、身份证'],
  otherGuardians: ['其他监护人'],
  familyEconomy: ['家庭经济'],
  address: ['家庭地址'],
  admitPersons: ['入住人']
}

// Helpers for Chinese ID validation and similarity
function normId(s) {
  if (!s) return null
  const t = String(s).trim().toUpperCase().replace(/\s+/g, '')
  return t || null
}
function isValidId18(id) {
  id = normId(id)
  if (!id || id.length !== 18) return false
  if (!/^\d{17}[0-9X]$/.test(id)) return false
  const weights = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
  const codes = ['1','0','X','9','8','7','6','5','4','3','2']
  let sum = 0
  for (let i = 0; i < 17; i++) sum += parseInt(id[i], 10) * weights[i]
  const check = codes[sum % 11]
  return check === id[17]
}
function hamming18(a, b) {
  a = normId(a); b = normId(b)
  if (!a || !b || a.length !== 18 || b.length !== 18) return Infinity
  let d = 0
  for (let i = 0; i < 18; i++) if (a[i] !== b[i]) d++
  return d
}

function toISO(d) {
  if (d === null || d === undefined || d === '') return null
  if (typeof d === 'number') {
    try {
      const x = XLSX.SSF.parse_date_code(d)
      return new Date(Date.UTC(x.y, x.m - 1, x.d)).toISOString().slice(0, 10)
    } catch (e) {}
  }
  const t = new Date(d)
  if (!isNaN(t.getTime())) return t.toISOString().slice(0, 10)
  // try replace dots with dashes
  const s = String(d).replace(/\./g, '-')
  const t2 = new Date(s)
  if (!isNaN(t2.getTime())) return t2.toISOString().slice(0, 10)
  return null
}

function pickPhoneAndId(raw) {
  raw = String(raw || '')
  const phoneMatch = raw.match(/1[3-9]\d{9}/)
  const idMatch = raw.match(/\d{17}[\dXx]/)
  let name = raw
  if (phoneMatch) name = name.replace(phoneMatch[0], '')
  if (idMatch) name = name.replace(idMatch[0], '')
  name = name.replace(/[\s，,、]+/g, ' ').trim()
  return {
    name: name || null,
    phone: phoneMatch ? phoneMatch[0] : null,
    id_card: idMatch ? idMatch[0] : null
  }
}

function mergeHeaderRows(ws) {
  const rows = XLSX.utils.sheet_to_json(ws, { header: 1, raw: true })
  const h1 = (rows[0] || []).map(h => (h == null ? null : String(h).trim()))
  const h2 = (rows[1] || []).map(h => (h == null ? null : String(h).trim()))
  const headers = []
  for (let i = 0; i < Math.max(h1.length, h2.length); i++) headers.push(h2[i] || h1[i] || `col_${i}`)
  const data = []
  for (let r = 2; r < rows.length; r++) {
    const arr = rows[r] || []
    if (!arr || arr.length === 0 || arr.every(x => x == null || String(x).trim() === '')) continue
    const o = {}
    for (let c = 0; c < headers.length; c++) o[headers[c]] = arr[c] == null ? '' : arr[c]
    data.push(o)
  }
  return { headers, rows: data }
}

function pick(row, keys) {
  for (const k of keys) {
    if (k in row) return row[k]
  }
  return ''
}

function tail4(idc) {
  const s = String(idc || '').replace(/\s/g, '')
  return s.length >= 4 ? s.slice(-4) : null
}

function norm(s) { return String(s || '').trim() || null }

function main() {
  const wb = XLSX.readFile(SRC)
  const ws = wb.Sheets[wb.SheetNames[0]]
  const merged = mergeHeaderRows(ws)
  const rows = merged.rows

  // Aggregation maps
  const patientsById = new Map() // id_card -> patient
  const patientsByName = new Map() // name -> patient (for records without id only)
  const patientsArr = []
  const tenanciesArr = []
  const tenKeySet = new Set()

  rows.forEach((r, idx) => {
    const name = norm(pick(r, COLS.name))
    let idc = normId(pick(r, COLS.id_card))
    const cin = toISO(pick(r, COLS.checkInDate))
    const birth = toISO(pick(r, COLS.birthDate))
    // Extract from parents if missing
    if (!idc) {
      const f = pickPhoneAndId(norm(pick(r, COLS.fatherMixed)) || '')
      const m = pickPhoneAndId(norm(pick(r, COLS.motherMixed)) || '')
      idc = normId(m.id_card || f.id_card || null)
    }
    // Parent info (for identity resolution)
    const gender = norm(pick(r, COLS.gender))
    const nativePlace = norm(pick(r, COLS.nativePlace))
    const ethnicity = norm(pick(r, COLS.ethnicity))
    const address = norm(pick(r, COLS.address))
    const hospital = norm(pick(r, COLS.hospital))
    const hospitalDiagnosis = norm(pick(r, COLS.hospitalDiagnosis))
    const doctorName = norm(pick(r, COLS.doctorName))
    const symptoms = norm(pick(r, COLS.symptoms))
    const medicalCourse = norm(pick(r, COLS.medicalCourse))
    const followupPlan = norm(pick(r, COLS.followupPlan))
    const father = pickPhoneAndId(norm(pick(r, COLS.fatherMixed)) || '')
    const mother = pickPhoneAndId(norm(pick(r, COLS.motherMixed)) || '')
    father.id_card = normId(father.id_card)
    mother.id_card = normId(mother.id_card)
    const otherGuardians = norm(pick(r, COLS.otherGuardians))
    const familyEconomy = norm(pick(r, COLS.familyEconomy))
    const admitPersons = norm(pick(r, COLS.admitPersons))

    // Resolve/Create patient per enhanced rule set
    // 1) If id equals a parent's id -> treat as child's id missing
    if (idc && (idc === father.id_card || idc === mother.id_card)) {
      idc = null
    }
    // 2) If id present but invalid checksum, try fuzzy match existing ids (<=2 chars diff) with same name+birth/phone
    const primaryPhone = mother.phone || father.phone || null
    if (idc && !isValidId18(idc)) {
      let mapped = null
      if (patientsById.size > 0 && (name || birth || primaryPhone)) {
        for (const [eid, ep] of patientsById.entries()) {
          const d = hamming18(idc, eid)
          if (d <= 2) {
            const sameName = !name || (ep.name && ep.name === name)
            const sameBirth = !birth || (ep.birthDate && ep.birthDate === birth)
            const phoneOverlap = primaryPhone && ((ep.motherPhone && ep.motherPhone === primaryPhone) || (ep.fatherPhone && ep.fatherPhone === primaryPhone))
            if ((sameName && sameBirth) || phoneOverlap) { mapped = eid; break }
          }
        }
      }
      if (mapped) idc = mapped; else idc = null
    }

    let p = null
    // 3) Try exact id match
    if (idc && patientsById.has(idc)) p = patientsById.get(idc)
    // 4) Without id, try name + birth + primaryPhone
    const nbpKey = name && birth && primaryPhone ? `${name}|${birth}|${primaryPhone}` : null
    if (!p && nbpKey && patientsByName.get(nbpKey)) p = patientsByName.get(nbpKey)
    // 5) Fallback: name + birth
    const nbKey = name && birth ? `${name}|${birth}` : null
    if (!p && nbKey && patientsByName.get(nbKey)) p = patientsByName.get(nbKey)
    // 6) Fallback: name only
    if (!p && name) p = patientsByName.get(name) || null

    const shouldCreateNew = () => {
      if (!p) return true
      // If both have id and different, but looks like typo (<=2 diffs) and other features align, then NOT a new patient
      if (p.id_card && idc && p.id_card !== idc) {
        const typo = hamming18(p.id_card, idc) <= 2
        const sameName = !name || (p.name === name)
        const sameBirth = !birth || (p.birthDate === birth)
        const phoneOverlap = primaryPhone && ((p.motherPhone === primaryPhone) || (p.fatherPhone === primaryPhone))
        if (typo && (sameName || sameBirth || phoneOverlap)) return false
        return true
      }
      return false
    }

    if (shouldCreateNew()) {
      p = {
        name: name || null,
        id_card: idc || null,
        id_card_tail: idc ? tail4(idc) : null,
        birthDate: birth,
        gender, nativePlace, ethnicity, address,
        hospital, hospitalDiagnosis, doctorName, symptoms, medicalCourse, followupPlan,
        fatherName: father.name, fatherPhone: father.phone, fatherIdCard: father.id_card,
        motherName: mother.name, motherPhone: mother.phone, motherIdCard: mother.id_card,
        otherGuardians, familyEconomy,
        createdAt: Date.now()
      }
      patientsArr.push(p)
      if (p.id_card) patientsById.set(p.id_card, p)
      if (nbpKey) patientsByName.set(nbpKey, p)
      else if (nbKey) patientsByName.set(nbKey, p)
      else if (p.name) patientsByName.set(p.name, p)
    } else {
      // Merge missing fields
      if (!p.id_card && idc) { p.id_card = idc; p.id_card_tail = tail4(idc); patientsById.set(idc, p) }
      const fill = (k, v) => { if (v && !p[k]) p[k] = v }
      fill('birthDate', birth)
      ;['gender','nativePlace','ethnicity','address','hospital','hospitalDiagnosis','doctorName','symptoms','medicalCourse','followupPlan','otherGuardians','familyEconomy'].forEach(k => fill(k, eval(k)))
      ;[['fatherName', father.name],['fatherPhone', father.phone],['fatherIdCard', father.id_card],['motherName', mother.name],['motherPhone', mother.phone],['motherIdCard', mother.id_card]].forEach(([k,v])=>fill(k,v))
    }

    // Tenancy
    if (cin || admitPersons) {
      const keyLeft = idc || (name || `ROW#${idx}`)
      const dupKey = `${keyLeft}|${cin || ''}|${(admitPersons||'').replace(/\s+/g,'')}`
      if (!tenKeySet.has(dupKey)) {
        tenKeySet.add(dupKey)
        tenanciesArr.push({
          patientKey: keyLeft,
          id_card: idc || null,
          patientName: name || null,
          checkInDate: cin,
          checkOutDate: null,
          room: null,
          bed: null,
          subsidy: null,
          extra: { admitPersons },
          createdAt: Date.now()
        })
      }
    }
  })

  // Post-merge pass: collapse same-name records where one has id and another doesn't
  const byName = new Map()
  patientsArr.forEach((p, i) => {
    const n = (p.name || '').trim()
    if (!n) return
    if (!byName.has(n)) byName.set(n, [])
    byName.get(n).push({ p, i })
  })
  const toRemove = new Set()
  for (const [n, items] of byName.entries()) {
    if (items.length < 2) continue
    const withId = items.filter(x => x.p.id_card)
    const withoutId = items.filter(x => !x.p.id_card)
    if (withId.length >= 1 && withoutId.length >= 1) {
      const base = withId[0].p
      for (const x of withoutId) {
        const q = x.p
        const fill = (k, v) => { if (v && !base[k]) base[k] = v }
        ;['birthDate','gender','nativePlace','ethnicity','address','hospital','hospitalDiagnosis','doctorName','symptoms','medicalCourse','followupPlan','otherGuardians','familyEconomy'].forEach(k => fill(k, q[k]))
        ;[['fatherName','fatherPhone','fatherIdCard'],['motherName','motherPhone','motherIdCard']].forEach(keys => keys.forEach(k => fill(k, q[k])))
        toRemove.add(x.i)
      }
    }
  }
  const patientsMerged = patientsArr.filter((_, idx) => !toRemove.has(idx))

  // Write JSONL files with .json suffix
  if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true })
  const fPatients = path.join(OUT_DIR, 'patients.json')
  const fTenancies = path.join(OUT_DIR, 'tenancies.json')
  const toLines = arr => arr.map(o => JSON.stringify(o)).join('\n') + '\n'
  fs.writeFileSync(fPatients, toLines(patientsMerged), 'utf8')
  fs.writeFileSync(fTenancies, toLines(tenanciesArr), 'utf8')

  const summary = {
    src: SRC,
    out: { patients: fPatients, tenancies: fTenancies },
    counts: { patients: patientsMerged.length, tenancies: tenanciesArr.length }
  }
  console.log(JSON.stringify(summary, null, 2))
}

main()

File: scripts/export_jsonl.py
#!/usr/bin/env python3
import json, os, re, sys, subprocess, tempfile
from datetime import datetime, timedelta

SRC = sys.argv[1] if len(sys.argv) > 1 else os.path.join(os.getcwd(), 'prepare', 'b.xlsx')
OUT_DIR = sys.argv[2] if len(sys.argv) > 2 else os.path.join(os.getcwd(), 'output')

COLS = {
  'name': ['姓名'],
  'id_card': ['身份证号'],
  'checkInDate': ['入住时间'],
  'birthDate': ['出生日期'],
  'gender': ['性别'],
  'nativePlace': ['籍贯'],
  'ethnicity': ['民族'],
  'hospital': ['就诊医院'],
  'hospitalDiagnosis': ['医院诊断'],
  'doctorName': ['医生姓名'],
  'symptoms': ['症状详情'],
  'medicalCourse': ['医治过程','医疗过程'],
  'followupPlan': ['后续治疗安排'],
  'fatherMixed': ['父亲姓名、电话、身份证号'],
  'motherMixed': ['母亲姓名、电话、身份证号','母亲姓名、电话、身份证'],
  'otherGuardians': ['其他监护人'],
  'familyEconomy': ['家庭经济'],
  'address': ['家庭地址'],
  'admitPersons': ['入住人']
}

def norm(s):
  if s is None: return None
  s = str(s).strip()
  return s if s else None

def norm_id(s):
  s = norm(s)
  if not s: return None
  return s.upper().replace(' ', '')

def is_valid_id18(idc):
  idc = norm_id(idc)
  if not idc or len(idc) != 18: return False
  if not re.match(r'^\d{17}[0-9X]$', idc): return False
  weights = [7,9,10,5,8,4,2,1,6,3,7,9,10,5,8,4,2]
  codes = ['1','0','X','9','8','7','6','5','4','3','2']
  s = sum(int(idc[i]) * weights[i] for i in range(17))
  return codes[s % 11] == idc[17]

def hamming18(a, b):
  a = norm_id(a); b = norm_id(b)
  if not a or not b or len(a)!=18 or len(b)!=18: return 999
  return sum(1 for i in range(18) if a[i]!=b[i])

def to_iso(d):
  if d is None or d == '':
    return None
  if isinstance(d, (int, float)):
    try:
      base = datetime(1899,12,30)
      dt = base + timedelta(days=float(d))
      return dt.strftime('%Y-%m-%d')
    except Exception:
      pass
  s = str(d).strip()
  if not s: return None
  s2 = s.replace('.', '-')
  for cand in (s, s2):
    try:
      dt = datetime.fromisoformat(cand)
      return dt.strftime('%Y-%m-%d')
    except Exception:
      pass
    try:
      dt = datetime.strptime(cand, '%Y-%m-%d')
      return dt.strftime('%Y-%m-%d')
    except Exception:
      pass
  return None

def pick_phone_id(raw):
  raw = norm(raw) or ''
  mphone = re.search(r'1[3-9]\d{9}', raw)
  mid = re.search(r'\d{17}[\dXx]', raw)
  name = raw
  if mphone: name = name.replace(mphone.group(0), '')
  if mid: name = name.replace(mid.group(0), '')
  name = re.sub(r'[\s，,、]+', ' ', name).strip()
  return {
    'name': name or None,
    'phone': mphone.group(0) if mphone else None,
    'id_card': norm_id(mid.group(0)) if mid else None
  }


def node_dump_rows(xlsx_path):
  js = ("""
const path=require('path');
let XLSX; try { XLSX=require('xlsx') } catch(e) { XLSX=require(path.resolve('functions/import-xlsx/node_modules/xlsx')) }
const p=__PATH__;
const wb=XLSX.readFile(p);
const ws=wb.Sheets[wb.SheetNames[0]];
const rows=XLSX.utils.sheet_to_json(ws,{header:1,raw:true});
const h1=(rows[0]||[]).map(h=>h==null?null:String(h).trim());
const h2=(rows[1]||[]).map(h=>h==null?null:String(h).trim());
const headers=[]; for(let i=0;i<Math.max(h1.length,h2.length);i++){headers.push(h2[i]||h1[i]||`col_${i}`)}
for(let r=2;r<rows.length;r++){
  const arr=rows[r]||[];
  if(!arr || arr.length===0 || arr.every(x=>x==null || String(x).trim()==='')) continue;
  const o={}; for(let c=0;c<headers.length;c++){o[headers[c]]=(arr[c]==null?'':arr[c]);}
  console.log(JSON.stringify(o));
}
""")
  js = js.replace('__PATH__', json.dumps(xlsx_path))
  cmd = ['node','-e',js]
  out = subprocess.check_output(cmd, text=True)
  return [json.loads(line) for line in out.strip().splitlines() if line.strip()]


def pick(row, keys):
  for k in keys:
    if k in row: return row[k]
  return ''

def tail4(idc):
  s = (norm_id(idc) or '')
  return s[-4:] if len(s)>=4 else None

def main():
  rows = node_dump_rows(SRC)

  patients_by_id = {}
  patients_by_key = {}
  patients = []
  tenancies = []
  ten_keys = set()

  for idx, r in enumerate(rows):
    name = norm(pick(r, COLS['name']))
    idc = norm_id(pick(r, COLS['id_card']))
    cin = to_iso(pick(r, COLS['checkInDate']))
    birth = to_iso(pick(r, COLS['birthDate']))
    gender = norm(pick(r, COLS['gender']))
    nativePlace = norm(pick(r, COLS['nativePlace']))
    ethnicity = norm(pick(r, COLS['ethnicity']))
    address = norm(pick(r, COLS['address']))
    hospital = norm(pick(r, COLS['hospital']))
    hospitalDiagnosis = norm(pick(r, COLS['hospitalDiagnosis']))
    doctorName = norm(pick(r, COLS['doctorName']))
    symptoms = norm(pick(r, COLS['symptoms']))
    medicalCourse = norm(pick(r, COLS['medicalCourse']))
    followupPlan = norm(pick(r, COLS['followupPlan']))
    father = pick_phone_id(pick(r, COLS['fatherMixed']))
    mother = pick_phone_id(pick(r, COLS['motherMixed']))
    father['id_card'] = norm_id(father['id_card'])
    mother['id_card'] = norm_id(mother['id_card'])
    otherGuardians = norm(pick(r, COLS['otherGuardians']))
    familyEconomy = norm(pick(r, COLS['familyEconomy']))
    admitPersons = norm(pick(r, COLS['admitPersons']))

    if idc and (idc == father['id_card'] or idc == mother['id_card']):
      idc = None

    primary_phone = mother['phone'] or father['phone']
    if idc and not is_valid_id18(idc):
      mapped = None
      for eid, ep in patients_by_id.items():
        if hamming18(idc, eid) <= 2:
          sameName = (not name) or (ep.get('name') == name)
          sameBirth = (not birth) or (ep.get('birthDate') == birth)
          phoneOverlap = primary_phone and (ep.get('motherPhone') == primary_phone or ep.get('fatherPhone') == primary_phone)
          if (sameName and sameBirth) or phoneOverlap:
            mapped = eid
            break
      idc = mapped

    p = None
    if idc and idc in patients_by_id:
      p = patients_by_id[idc]
    if not p:
      keys = []
      if name and birth and primary_phone:
        keys.append(('nbp', f'{name}|{birth}|{primary_phone}'))
      if name and birth:
        keys.append(('nb', f'{name}|{birth}'))
      if name:
        keys.append(('n', name))
      for k in keys:
        if k in patients_by_key:
          p = patients_by_key[k]
          break

    create_new = False
    if not p:
      create_new = True
    else:
      if p.get('id_card') and idc and p['id_card'] != idc:
        typo = hamming18(p['id_card'], idc) <= 2
        sameName = (not name) or (p.get('name') == name)
        sameBirth = (not birth) or (p.get('birthDate') == birth)
        phoneOverlap = primary_phone and (p.get('motherPhone') == primary_phone or p.get('fatherPhone') == primary_phone)
        if not (typo and (sameName or sameBirth or phoneOverlap)):
          create_new = True

    # Decide if this row has enough info to create a patient
    has_identity = bool(name or idc)
    has_profile = any([
      birth, gender, nativePlace, ethnicity, address, hospital, hospitalDiagnosis,
      doctorName, symptoms, medicalCourse, followupPlan,
      father['name'], father['phone'], father['id_card'],
      mother['name'], mother['phone'], mother['id_card'],
      otherGuardians, familyEconomy
    ])

    if create_new and (has_identity or has_profile):
      p = {
        'name': name,
        'id_card': idc,
        'id_card_tail': tail4(idc),
        'birthDate': birth,
        'gender': gender,
        'nativePlace': nativePlace,
        'ethnicity': ethnicity,
        'address': address,
        'hospital': hospital,
        'hospitalDiagnosis': hospitalDiagnosis,
        'doctorName': doctorName,
        'symptoms': symptoms,
        'medicalCourse': medicalCourse,
        'followupPlan': followupPlan,
        'fatherName': father['name'], 'fatherPhone': father['phone'], 'fatherIdCard': father['id_card'],
        'motherName': mother['name'], 'motherPhone': mother['phone'], 'motherIdCard': mother['id_card'],
        'otherGuardians': otherGuardians,
        'familyEconomy': familyEconomy,
        'createdAt': int(datetime.utcnow().timestamp()*1000)
      }
      patients.append(p)
      if idc: patients_by_id[idc] = p
      if name and birth and primary_phone: patients_by_key[('nbp', f'{name}|{birth}|{primary_phone}')] = p
      elif name and birth: patients_by_key[('nb', f'{name}|{birth}')] = p
      elif name: patients_by_key[('n', name)] = p
    elif not create_new and p is not None:
      def fill(k,v):
        if v and not p.get(k): p[k]=v
      fill('birthDate', birth)
      for k,v in [('gender',gender),('nativePlace',nativePlace),('ethnicity',ethnicity),('address',address),('hospital',hospital),('hospitalDiagnosis',hospitalDiagnosis),('doctorName',doctorName),('symptoms',symptoms),('medicalCourse',medicalCourse),('followupPlan',followupPlan),('otherGuardians',otherGuardians),('familyEconomy',familyEconomy)]:
        fill(k,v)
      for k,v in [('fatherName',father['name']),('fatherPhone',father['phone']),('fatherIdCard',father['id_card']),('motherName',mother['name']),('motherPhone',mother['phone']),('motherIdCard',mother['id_card'])]:
        fill(k,v)
      if not p.get('id_card') and idc:
        p['id_card']=idc; p['id_card_tail']=tail4(idc); patients_by_id[idc]=p

    if cin or admitPersons:
      keyLeft = idc or (name or f'ROW#{idx}')
      dupKey = f"{keyLeft}|{cin or ''}|{(admitPersons or '').replace(' ','')}"
      if dupKey not in ten_keys:
        ten_keys.add(dupKey)
        tenancies.append({
          'patientKey': keyLeft,
          'id_card': idc,
          'patientName': name,
          'checkInDate': cin,
          'checkOutDate': None,
          'room': None,
          'bed': None,
          'subsidy': None,
          'extra': { 'admitPersons': admitPersons },
          'createdAt': int(datetime.utcnow().timestamp()*1000)
        })

  by_name = {}
  for i,p in enumerate(patients):
    n = (p.get('name') or '').strip()
    if not n: continue
    by_name.setdefault(n, []).append((i,p))
  remove_idx = set()
  for n, items in by_name.items():
    with_id = [x for x in items if x[1].get('id_card')]
    without_id = [x for x in items if not x[1].get('id_card')]
    if with_id and without_id:
      base = with_id[0][1]
      for i,q in without_id:
        def fill(k,v):
          if v and not base.get(k): base[k]=v
        fill('birthDate', q.get('birthDate'))
        for k in ['gender','nativePlace','ethnicity','address','hospital','hospitalDiagnosis','doctorName','symptoms','medicalCourse','followupPlan','otherGuardians','familyEconomy']:
          fill(k, q.get(k))
        for k in ['fatherName','fatherPhone','fatherIdCard','motherName','motherPhone','motherIdCard']:
          fill(k, q.get(k))
        remove_idx.add(i)
  patients = [p for i,p in enumerate(patients) if i not in remove_idx]

  # Final filter: drop records that are entirely empty (no name, no id, and no core fields)
  def is_meaningful(p):
    if p.get('name') or p.get('id_card'):
      return True
    keys = ['birthDate','gender','nativePlace','ethnicity','address','hospital','hospitalDiagnosis','doctorName','symptoms','medicalCourse','followupPlan','fatherName','fatherPhone','fatherIdCard','motherName','motherPhone','motherIdCard','otherGuardians','familyEconomy']
    return any(p.get(k) for k in keys)
  patients = [p for p in patients if is_meaningful(p)]

  os.makedirs(OUT_DIR, exist_ok=True)
  f_pat = os.path.join(OUT_DIR, 'patients.json')
  f_ten = os.path.join(OUT_DIR, 'tenancies.json')
  with open(f_pat,'w',encoding='utf-8') as f:
    for p in patients: f.write(json.dumps(p, ensure_ascii=False)+'\n')
  with open(f_ten,'w',encoding='utf-8') as f:
    for t in tenancies: f.write(json.dumps(t, ensure_ascii=False)+'\n')

  print(json.dumps({ 'src': SRC, 'out': { 'patients': f_pat, 'tenancies': f_ten }, 'counts': { 'patients': len(patients), 'tenancies': len(tenancies) } }, ensure_ascii=False))

if __name__ == '__main__':
  main()
